File: python/txtvm/op.py
from __future__ import absolute_import as _abs

__binary_op_cls = None

class BinaryOp(object):
    """Base class of binary operator"""
    def __call__(self, lhs, rhs):
        return __binary_op_cls(self, lhs, rhs)

class AddOp(BinaryOp):
    pass

class SubOp(BinaryOp):
    pass

class MulOp(BinaryOp):
    pass

class DivOp(BinaryOp):
    pass

add = AddOp()
sub = SubOp()
mul = MulOp()
div = DivOp()

File: src/expr/op.cc
#include <txtvm/op.h>
#include <txtvm/expr_node.h>

namespace dmlc {
DMLC_REGISTRY_ENABLE(::tvm::BinaryOpReg);
DMLC_REGISTRY_ENABLE(::tvm::UnaryOpReg);
}; // end of namespace dmlc

namespace tvm {

Expr BinaryOp::operator()(Expr lhs, Expr rhs) const {
    auto nptr = std::make_shared<BinaryOpNode>(
            this, std::move(lhs), std::move(rhs));
    nptr->Verify();
    return Expr(std::move(nptr));
}

Expr BinaryOp::Reduce(Expr src, RDomain rdom) const {
    auto nptr = std::make_shared<ReduceNode>(
        this, std::move(src), std::move(rdom));
    nptr->Verify();
    return Expr(std::move(nptr));
}

const BinaryOp* BinaryOp::Get(const char* name) {
    const auto* op = dmlc::Registry<BinaryOpReg>::Find(name);
    CHECK(op != nullptr) << "cannot find " << name;
    return op->op.get();
}

TVM_REGISTER_BINARY_OP(+, AddOp);
TVM_REGISTER_BINARY_OP(-, SubOp);
TVM_REGISTER_BINARY_OP(*, MulOp);
TVM_REGISTER_BINARY_OP(/, DivOp);
TVM_REGISTER_BINARY_OP(max, MaxOp);
TVM_REGISTER_BINARY_OP(min, MinOp);

} // end of namespace tvm
